nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java

```diff
@@ -0,0 +1,387 @@
+ * where filename is the property name and file contents is the property value.
+ * @since 5.2020.7
+ */
+public class DirConfigSource extends PayaraConfigSource implements ConfigSource {
+
+    static final class DirProperty {
+        final String propertyValue;
+        final FileTime lastModifiedTime;
+        final Path path;
+        final int pathDepth;
+        
+        DirProperty(String propertyValue, FileTime lastModifiedTime, Path path, Path rootPath) {
+            this.propertyValue = propertyValue;
+            this.lastModifiedTime = lastModifiedTime;
+            this.path = path;
+            this.pathDepth = rootPath.relativize(path).getNameCount();
+        }
+    
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            DirProperty that = (DirProperty) o;
+            return pathDepth == that.pathDepth &&
+                propertyValue.equals(that.propertyValue) &&
+                lastModifiedTime.equals(that.lastModifiedTime) &&
+                path.equals(that.path);
+        }
+    
+        @Override
+        public int hashCode() {
+            return Objects.hash(propertyValue, lastModifiedTime, path, pathDepth);
+        }
+    }
+    
+    final class DirPropertyWatcher implements Runnable {
+    
+        private final Logger logger = Logger.getLogger(DirConfigSource.class.getName());
+        private final WatchService watcher = FileSystems.getDefault().newWatchService();
+        private final ConcurrentHashMap<WatchKey, Path> watchedFileKeys = new ConcurrentHashMap<>();
+        
+        DirPropertyWatcher(Path topmostDir) throws IOException {
+            if (Files.exists(topmostDir) && Files.isDirectory(topmostDir) && Files.isReadable(topmostDir)) {
+                registerAll(topmostDir);
+            } else {
+                throw new IOException("Given directory '"+topmostDir+"' is no directory or cannot be read.");
+            }
+        }
+        
+        void registerAll(Path dir) throws IOException {
+            // register file watchers recursively (they don't attach to subdirs...)
+            Files.walkFileTree(dir, new SimpleFileVisitor<Path>() {
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException
+                {
+                    register(dir);
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        }
+        
+        void register(Path dir) throws IOException {
+            WatchKey key = dir.register(watcher, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+            watchedFileKeys.putIfAbsent(key, dir);
+        }
+        
+        @Override
+        public void run() {
+            while (true) {
+                // wait infinitely until we receive an event (or the executor is shutting down)
```

Comment:
`PayaraExecutorService` tries to terminate gracefully via `awaitTermination` first. This loop therefore delays shutdown of the server by 5 seconds.
Consider scheduling single `take` every few seconds instead, or polling synchronously during get.

Reply:
This issue has been taken care of.